HoRNDIS.cpp

```diff
@@ -582,182 +417,297 @@ bool HoRNDIS::createNetworkInterface() {
 }
 
 /***** Interface enable and disable logic *****/
+HoRNDIS::ReentryLocker::ReentryLocker(IOCommandGate *inGate, bool &inGuard)
+		: gate(inGate), entryGuard(inGuard), result(kIOReturnSuccess) {
+	// Wait until we exit the previously-entered enable or disable method:
+	while (entryGuard) {
+		LOG(V_DEBUG, "Delaying the re-entered call");
+		result = gate->commandSleep(&entryGuard);
+		// If "commandSleep" has failed, stop immediately, and don't
+		// touch the 'entryGuard':
+		if (isInterrupted()) {
+			return;
+		}
+	}
+	entryGuard = true;  // Mark the entry into one of the protected methods.
+}
 
-/* Contains buffer alloc and dealloc, notably.  Why do that here?  Because that's what Apple did. */
+HoRNDIS::ReentryLocker::~ReentryLocker() {
+	if (!isInterrupted()) {
+		entryGuard = false;
+		gate->commandWakeup(&entryGuard);
+	}
+}
 
+/*!
+ * Calls 'pipe->io', and if there is a stall, tries to clear that 
+ * stall and calls it again.
+ */
+static inline IOReturn robustIO(IOUSBHostPipe* pipe, pipebuf_t *buf,
+	uint32_t len) {
+	IOReturn rc = pipe->io(buf->mdp, len, &buf->comp);
+	if (rc == kUSBHostReturnPipeStalled) {
+		LOG(V_DEBUG, "USB Input Pipe is stalled. Clearing and re-trying");
+		rc = pipe->clearStall(true);
+		if (rc != kIOReturnSuccess) {
+			LOG(V_ERROR, "Failed to clear pipe stall: %08x", rc);
+			// Don't fail if 'clearStall' does not work.
+		}
+		rc = pipe->io(buf->mdp, len, &buf->comp);
+	}
+	return rc;
+}
+
+/* Contains buffer alloc and dealloc, notably.  Why do that here?  
+   Not just because that's what Apple did. We don't want to consume these 
+   resources when the interface is sitting disabled and unused. */
```

Comment:
Somebody has to care about those poor Macs with only 4GB of RAM!  Apple sure doesn't ...

Reply:
But we can be proud of having such a high-efficiency driver! :)
Seriously, you're right. Plus, no one ever shall be running the driver in the disabled state - it will either be enabled, or fully unloaded.